src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java

```diff
@@ -338,20 +342,37 @@ public String uri() {
     return uri;
   }
 
+  
   @Override
   public String path() {
-    synchronized (conn) {
-      this.path = uri != null ? HttpUtils.parsePath(uri) : null;
-      return path;
+    if (path == null) {
+      synchronized (conn) {
+        if (path == null) {
+          final Object o =  HttpUtils.parsePathAndQueryStartIf(uri());
+          if (o instanceof Object[]) {
+            final Object[] arr =  (Object[])o;
+            path = (String)arr[0];
+            queryStart = (Integer)arr[1];
+          }else {
+            path = (String)o;
+          }
+         }
+       }
     }
+    return path;
   }
 
   @Override
   public String query() {
-    synchronized (conn) {
-      this.query = uri != null ? HttpUtils.parseQuery(uri) : null;
-      return query;
+    if (query == null) {
+      synchronized (conn) {
+        if (query == null) {
+          query = path() == uri || queryStart == -1 ? null : queryStart > 0 && uri.length() > queryStart
```

Comment:
This becomes unreadable with so many chain of logic operations and ternary blocks. I believe proper if statements would be preferred and more readable

Reply:
if path already init( path != null ), should we "synchronized (conn) " for #path ?